### ISSUE TITLE
Updates Needed for Ema and Orange Displays

### ISSUE THREAD

**Issue by marionbarker** [ps2/rileylink_ios#686]
This comment was updated (Oct 12, 2021) to include vs2 of the patch and with easier usage instructions (Oct 16, 2021).

This comment was updated on July 5, 2022 with the following changes:
* This code was added to dev but not master
  * Loop 2.2.9 needs the patch to see EmaLink `Battery level` or Orange Link Pro `Find Device` in the RileyLink display.
* The rileylink_ios commit level for a fresh download is now 1651753 - and patch still works. 
  * I did not rename the patch, but have updated the instructions where it says, must be at 15d1997, to now say: must be at 15d1997 or 1651753.
* I have removed references to running this on dev because no patch is needed with dev.

## Purpose of Patch:

The items displayed in the RileyLink Display need some updates to handle:
* OrangeLink and OrangeLink Pro
  * Interpret hardware type for all released versions of OL and OL Pro firmware
  * Show the "Find Device" line for HW1.1 (OL Pro) regardless of firmware
* EmaLink
  * Method to identify EmaLink
  * Battery display
  * Add battery alert level picker
  * Change label to "Invert LED Logic" (actual function of "Diagnostic LEDs" row for .riley)

The battery alert picker sets a common level for all .ema and .orange
* If you switch between these two, you may choose to adjust the level alert for whichever device is currently selected

## Step 0: Open Terminal at LoopWorkspace Folder

For each command example shown below, there is a copy button that appears when you hover your mouse to the upper right corner of the example.  You can tap on the copy button and then paste it into your terminal.

You must in be the LoopWorkspace folder to use these commands. Each set of commands, in later steps, returns you to LoopWorkspace.  (If there is an error, you probably need to start over.)

If you are running dev (see notes below)

Open a terminal window with path at the LoopWorkspace folder used for the build. 

You can use Finder to do this (these instructions assume you used the Build-Select Script - if your files are in a different folder, make the appropriate adjustment)
1. Use Finder to navigate to Downloads / BuildLoop and open the folder with the most recent date (e.g., Loop-Master-211006-0524)
1. Right click (or hold down control key and click) on LoopWorkspace folder
1. In drop-down menu, select New Terminal at Folder

This opens a new terminal in the correct folder.  You will use this terminal window for Steps 1 - 3.

But first, make sure you are in the correct directory, copy and paste this command into the terminal.

```
pwd
```

On my computer, it shows: `/Users/marion/Downloads/BuildLoop/Loop-Master-211006-0524/LoopWorkspace`

If your response does not end in LoopWorkspace, you are not in the correct directory, which means the commands in the next steps will not work as documented.

Option 1: Quit out of terminal and try again
Option 2: If you are one directory too far, e.g., 'LoopWorkspace/Loop', type `cd ..` and return

## Step 1: Check Status

Make sure you have the correct version and that there are no changes to your rileylink_ios folder.

```
cd rileylink_ios; git status; cd ..
```

If the status shows
```
HEAD detached at 15d1997
nothing to commit, working tree clean
```
or
```
HEAD detached at 1651753
nothing to commit, working tree clean
```

You are good to proceed. If it shows one or more files modified, you may have a different patch - you need to remove that first.  See Clear old Patches below. If it does not show 15d1997 or 1651753 DO NOT USE this patch - ask for help.

## Step 2: Apply Patch

Download the patch to your Downloads folder (that is the default location when you click on the patch link below.)
* [ema_ol_fix_master_15d1997_vs2.txt](https://github.com/ps2/rileylink_ios/files/7331334/ema_ol_fix_master_15d1997_vs2.txt)

These commands apply the patch:

```
cd rileylink_ios
git apply ~/Downloads/ema_ol_fix_master_15d1997_vs2.txt
cd ..
```

If patch applied successfully, there should be no errors reported.
If there were errors, figure them out or ask for help.

## Step 3: Open Xcode

Open Xcode to build a new version on your phone:

```
xed .
```

Assuming the build was successful, check your device display to ensure you are seeing the display expected.

## Clear old Patches:

Only use this if your HEAD was 15d1997 or 1651753.  Anything else, you can't use this patch.

To clear an old patch, issue these commands then return to Step 1:
```
cd rileylink_ios; git stash; cd ..
```

## Modified files:

This patch modifies the following 2 files:
* RileyLinkBLEKit/RileyLinkDevice.swift
* RileyLinkKitUI/RileyLinkDeviceTableViewController.swift

## Details about vs2:

What does vs2 offer over the first version?
* It distinguishes between .riley and .ema
* It adds a battery alert picker to .ema

Please note:
* Assumptions:
  * the ble_rfspy version for .riley is 2.x
  * the ble_rfspy version for .ema is 3.x
    * If you are running RileyLink hardware with EmaLink firmware, this patch treats it like an EmaLink but the battery level is bogus because RL hardware doesn't have the component to read battery level.

## Reason for patch:

Updates are needed to handle Ema and Orange Displays for the LoopWorkspace build for master that points to rileylink_ios commit 15d1997 or 1651753.

The patch can be applied to the LoopWorkspace build for master (Loop v2.2.6) that points to rileylink_ios commit 15d1997 or 1651753. 

## People running dev

This patch is not required for dev.

**Comment by jlucasvt** [comment]
Marion. I'm running the latest L&L build script workspace and get an error when trying to apply the the patch.

**Comment by marionbarker** [comment]
Screenshot? 
* Are you getting a patch error or a build error?
* Are you using Xcode 13?
* Are you aware of this Xcode 13 build error and solution?
  * https://loopkit.github.io/loopdocs/build/build_errors/#xcode-13-indexing-not-finished-watchapp-extension-entitlements

I don't know why Xcode 13 is so difficult and throws this error so often.  
But that would be a different issue - not to get off topic here.

I suggest you build successfully before applying the patch.
For me, I had to clear the WatchApp Extension.entitlements error, then build.
Follow the steps as listed above and shown in this graphic (I added a few extra `git status` that are not required but confirm HEAD detached at 15d1997).
Successfully built with the patch and the Ema and Orange displays were updated as expected. 

![apply-patch-for-ema-ol](https://user-images.githubusercontent.com/19607791/135956001-305f827f-e25e-494f-9cf7-ce9564847c02.png)

Note - there are still some things about this patch that I think need adjusting. Happy to get feedback and suggestions.

**Comment by jlucasvt** [comment]
Yes I have XCode 13 and successfully build and deplyed to phone and watch.  I copied the patch to my rileylink_ios folder and attempted apply.

<img width="695" alt="Screen Shot 2021-10-05 at 8 42 06 AM" src="https://user-images.githubusercontent.com/19819388/136034901-de320415-21ba-469d-bbab-0cda42e7a30a.png">

**Comment by marionbarker** [comment]
You had already made changes to those 2 files.  You can only apply a patch to a fresh copy of rileylink_ios HEAD detached at 15d1997. (Or at least those two files cannot be modified from fresh copy.)

Try these commands to fix what you currently have. 
This moves the patch out of directory up one level, otherwise stash will remove it.

mv ema_ol_fix_master_15d1997.txt ..

Restore clean copy of rileylink_ios

git stash

Now apply the patch 

git apply ../ema_ol_fix_master_15d1997.txt

**Comment by jlucasvt** [comment]
Deerp. Thanks for my obvious dopey ness..   I've applied the patch and have deployed it to my device.  Testing now.

**Comment by jlucasvt** [comment]
Working great on my OL Pro.  Will do some further testing this week no problems thus far.

**Comment by sumairs** [comment]
Hi @marionbarker, I got a bunch of build errors when applying the patch to loop V2.2.6. I have an OL Pro FW3.2/HW1.1. 
Any idea what I may have done wrong?

<img width="1023" alt="Screen Shot 2022-01-05 at 6 24 54 PM" src="https://user-images.githubusercontent.com/57295742/148304245-272edecf-922c-4b51-81f2-2b51214b18c4.png">
<img width="1396" alt="Screen Shot 2022-01-05 at 6 22 28 PM" src="https://user-images.githubusercontent.com/57295742/148304249-f3f56e3a-e359-42ed-8f9d-aeee76e2dea6.png">

**Comment by ps2** [comment]
Are all of these changes in dev at this point?  If not, please identify what is remaining. If a patch can be created, I can review it, or I can look into getting the remaining issues resolved.

**Comment by marionbarker** [comment]
None of these changes are in dev yet.
Using this commit: 6321592:
* OrangeLink Pro does not display Find Device
* EmaLink does not display battery, battery level alert or Invert LED Logic row

I did 
```
git apply ema_ol_fix_master_15d1997_vs2.txt
```

and displays of both are as desired.

**Comment by marionbarker** [comment]
@ps2 I just looked at this again today based off a facebook comment.  I think there was some confusion when you closed this issue.  It's fixed in dev but not in master.  I've updated the patch instructions to include the updated commit number for rileylink_ios.  

We point to the [patch instructions in LoopDocs](https://loopkit.github.io/loopdocs/operation/loop-settings/rileylink/#rileylink-status-and-commands). Scroll down to the admonition that says `Patch Required for Loop 2.2.x`
